src/smt/optimization_solver.cpp

```diff
@@ -46,92 +46,371 @@ OptResult OptimizationSolver::checkOpt()
+            + aMod2PlusbMod2.logicalRightShift(bv1));
   }
+}
+
+std::unique_ptr<SmtEngine> OMTOptimizerBitVector::initOptChecker(
+    SmtEngine* parentSMTSolver)
+{
+  std::unique_ptr<SmtEngine> optChecker;
+  CVC4::theory::initializeSubsolver(optChecker);
+  // we need to be in incremental mode for multiple objectives since we need to
+  // push pop we need to produce models to inrement on our objective
+  optChecker->setOption("incremental", "true");
+  optChecker->setOption("produce-models", "true");
+  // Move assertions from the parent solver to the subsolver
+  std::vector<Node> p_assertions = parentSMTSolver->getExpandedAssertions();
+  for (const Node& e : p_assertions)
+  {
+    optChecker->assertFormula(e);
+  }
+  return optChecker;
+}
 
+std::pair<OptResult, Node> OMTOptimizerBitVector::minimize(
+    SmtEngine* parentSMTSolver, Node target)
+{
+  // the smt engine to which we send intermediate queries
+  // for the binary search.
+  std::unique_ptr<SmtEngine> optChecker = initOptChecker(parentSMTSolver);
+  NodeManager* nm = optChecker->getNodeManager();
+  Result intermediateSatResult = optChecker->checkSat();
   // Model-value of objective (used in optimization loop)
   Node value;
-  // asserts objective > old_value (used in optimization loop)
-  Node increment;
-
-  // Workhorse of linear optimization:
-  // This loop will keep incrmenting the objective until unsat
-  // When unsat is hit, the optimized value is the model value just before the
-  // unsat call
-  while (loop_r.isSat())
+  if (intermediateSatResult.isUnknown())
+  {
+    return std::make_pair(OptResult::OPT_UNKNOWN, value);
+  }
+  if (!intermediateSatResult.isSat())
   {
-    // get the model-value of objective in last sat call
-    value = optChecker->getValue(d_activatedObjective.getNode());
+    return std::make_pair(OptResult::OPT_UNSAT, value);
+  }
 
-    // We need to save the value since we need the model value just before the
-    // unsat call
-    Assert(!value.isNull());
-    d_savedValue = value;
+  // value equals to upperBound
+  value = optChecker->getValue(target);
+
+  // this gets the bitvector!
+  BitVector bvValue = value.getConst<BitVector>();
+  unsigned int bvSize = bvValue.getSize();
 
-    // increment on the model-value of objective:
-    // if we're maximizing increment = objective > old_objective value
-    // if we're minimizing increment = objective < old_objective value
-    if (d_activatedObjective.getType() == OBJECTIVE_MAXIMIZE)
+  // lowerbound
+  BitVector lowerBound = ((this->d_isSigned) ? (BitVector::mkMinSigned(bvSize))
+                                             : (BitVector::mkZero(bvSize)));
+  // upperbound must be a satisfying value
+  // and value == upperbound
+  BitVector upperBound = bvValue;
+
+  Kind LTOperator =
+      ((d_isSigned) ? (kind::BITVECTOR_SLT) : (kind::BITVECTOR_ULT));
+  Kind GEOperator =
+      ((d_isSigned) ? (kind::BITVECTOR_SGE) : (kind::BITVECTOR_UGE));
+  BitVector pivot;
```

Comment:
can u document what is the role of pivot above its declaration?

Reply:
hmm? good question, I don't know how to describe it......

Reply:
:) It's important in case someone ever modifies this code, to be able to understand. Maybe start by describing vaguely and we can see if we can improve? Doesn't have to be long, just something to put it in the right context.